desc/objectives/_coils.py

```
@@ -0,0 +1,504 @@
+        weight=1,
+        normalize=True,
+        normalize_target=True,
+        loss_function=None,
+        deriv_mode="auto",
+        grid=None,
+        name=None,
+    ):
+        self._grid = grid
+        self._data_keys = data_keys
+        self._normalize = normalize
+        super().__init__(
+            things=[coil],
+            target=target,
+            bounds=bounds,
+            weight=weight,
+            normalize=normalize,
+            normalize_target=normalize_target,
+            loss_function=loss_function,
+            deriv_mode=deriv_mode,
+            name=name,
+        )
+
+    def build(self, use_jit=True, verbose=1):
+        """Build constant arrays.
+
+        Parameters
+        ----------
+        use_jit : bool, optional
+            Whether to just-in-time compile the objective and derivatives.
+        verbose : int, optional
+            Level of output.
+
+        """
+        # local import to avoid circular import
+        from desc.coils import CoilSet, MixedCoilSet, _Coil
+
+        is_mixed_coils = isinstance(self.things[0], MixedCoilSet)
+        is_coil_set = isinstance(self.things[0], CoilSet)
+
+        coils = tree_flatten(
+            self.things[0],
+            is_leaf=lambda x: isinstance(x, _Coil) and not isinstance(x, CoilSet),
+        )[0]
+
+        # if using single coil, make coils and grid a list so they can be
+        # used with tree_map
+        coils = [coils[0]] if not is_coil_set else coils
+
+        if self._grid is None:
+            NFP = self.NFP if hasattr(self, "NFP") else 1
+            grid = lambda x: LinearGrid(N=2 * x + 5, NFP=NFP, endpoint=False)
+            self._grid = [grid(coil.N) for coil in coils]
+        elif self._grid.num_rho > 1 or self._grid.num_theta > 1:
+            raise TypeError("Only use toroidal resolution for coil grids.")
+        elif isinstance(self._grid, numbers.Integral):
+            raise TypeError("The inputted grid should be of type `Grid`.")
```

not sure if we need to throw an error for this? could just do
`self._grid = LinearGrid(N=self._grid)` in this case